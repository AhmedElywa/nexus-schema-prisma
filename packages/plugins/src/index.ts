export * from './select';
export * from './sdlInputs';
