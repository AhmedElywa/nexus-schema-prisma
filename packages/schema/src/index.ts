export * from '@paljs/types';
export * from './json';
export * from './camelCase';
export * from './typescript';
export * from './PrismaReader';
