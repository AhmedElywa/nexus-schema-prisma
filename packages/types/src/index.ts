export * from './cli';
export * from './generatedSchema';
export * from './generator';
export * from './configFile';
export * from './adminSchema';
export * from './dmmf';
