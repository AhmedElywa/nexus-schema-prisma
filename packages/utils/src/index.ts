export * from './dmmf';
export * from './schemaInputs';
export { getConfig, getEnvPaths, tryLoadEnvs, getDMMF } from '@prisma/internals';
