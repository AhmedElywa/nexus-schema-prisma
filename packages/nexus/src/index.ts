import { PrismaSelect } from '@paljs/plugins';
import type { DMMF } from '@paljs/types';
import { objectType, plugin } from 'nexus';
import { NexusAcceptedTypeDef } from 'nexus/dist/builder';
import { adminNexusSchemaSettings } from './admin';
import type { Settings } from './settings';
import { getScalars } from './defaultScalar';
import { AllNexusOutputTypeDefs } from 'nexus/dist/definitions/wrapping';

export { Settings };

export const paljs = (settings?: Settings) =>
  plugin({
    name: 'paljs',
    description:
      'paljs plugin to add Prisma select to your resolver and prisma admin queries and mutations and all models input types',
    onInstall(builder) {
      const nexusSchemaInputs: (NexusAcceptedTypeDef | AllNexusOutputTypeDefs)[] = [
        objectType({
          name: 'BatchPayload',
          definition(t) {
            t.nonNull.int('count');
          },
        }),
        ...getScalars(settings?.excludeScalar),
      ];

      if (settings?.includeAdmin) {
        nexusSchemaInputs.push(...adminNexusSchemaSettings(settings?.adminSchemaPath));
      }
      for (const type of nexusSchemaInputs) {
        builder.addType(type);
      }
    },
    onCreateFieldResolver() {
      return async (root, args, ctx, info: any, next) => {
        ctx.select = new PrismaSelect(info, {
          ...settings?.prismaSelectOptions,
          dmmf: settings?.prismaSelectOptions?.dmmf as Omit<DMMF.Document, 'schema'>[],
        });
        return await next(root, args, ctx, info);
      };
    },
  });
