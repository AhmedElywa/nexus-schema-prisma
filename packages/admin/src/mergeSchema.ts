import { Schema, SchemaModel, SchemaField } from './types';
import { SchemaObject, Model, Field } from '@prisma-tools/schema';

export function mergeSchema(object: SchemaObject, schema: Schema): Schema {
  const newSchema: Schema = {
    models: [],
    enums: object.enums,
  };
  object.models.forEach((item) => {
    const schemaItem = schema.models.find((model) => model.id === item.name);
    if (!schemaItem) {
      if (checkIdFieldExist(item)) {
        newSchema.models.push(handleNewModel(item));
      }
    } else {
      const newItem: SchemaModel = {
        ...schemaItem,
        idField:
          schemaItem.idField ?? item.fields.find((field) => field.isId)?.name,
        displayFields: schemaItem.displayFields ?? [
          item.fields.find((field) => field.isId)?.name,
        ],
        fields: [],
      };
      item.fields.forEach((field) => {
        const schemaField = schemaItem.fields.find(
          (item) => item.name === field.name,
        );
        if (!schemaField) {
          newItem.fields.push(handleNewField(field, schemaItem.name));
        } else {
          newItem.fields.push({
            ...schemaField,
            ...getOriginalField(field, schemaItem.name),
          });
        }
      });
      newItem.fields.sort((a, b) => a.order - b.order);
      newSchema.models.push(newItem);
    }
  });
  return newSchema;
}

function checkIdFieldExist(model: Model) {
  return !!model.fields.find((field) => field.isId);
}

function handleNewModel(model: Model) {
  const newItem: SchemaModel = {
    id: model.name,
    name: getTitle(model.name),
    idField: model.fields.find((field) => field.isId)!.name,
    displayFields: [model.fields.find((field) => field.isId)!.name],
    create: true,
    update: true,
    delete: true,
    fields: [],
  };
  model.fields.forEach((field) => {
    newItem.fields.push(handleNewField(field, model.name));
  });
  return newItem;
}

const defaultField = ['id', 'createdAt', 'updatedAt'];

function handleNewField(field: Field, modelName: string): SchemaField {
  return {
    ...getOriginalField(field, modelName),
    title: getTitle(field.name),
    create: !defaultField.includes(field.name) && !field.relationField,
    update: !defaultField.includes(field.name) && !field.relationField,
    read: true,
    filter: true,
    sort: true,
    order: -1,
  };
}

function getTitle(id: string) {
  const split = id.split(/(?=[A-Z])/);
  split[0] = split[0].charAt(0).toUpperCase() + split[0].slice(1);
  return split.join(' ');
}

function getOriginalField(
  field: Field,
  modelName: string,
): Omit<Field, 'relation'> & { id: string } {
  delete field.relation;
  return {
    id: modelName + '.' + field.name,
    ...field,
  };
}
