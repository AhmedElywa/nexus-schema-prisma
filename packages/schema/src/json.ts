import { SchemaObject, Model, Enums, Field } from '@paljs/types';
import { PrismaReader } from './PrismaReader';

export class ConvertSchemaToObject extends PrismaReader {
  schemaObject: SchemaObject = {
    models: [],
    enums: [],
  };

  constructor(protected path: string) {
    super(path);
  }

  run() {
    this.getModels();
    this.getEnums();
    return this.schemaObject;
  }

  private getModels() {
    if (this.models) {
      for (const model of this.models) {
        const lines = this.blockLines(model);
        const modelObject: Model = {
          name: this.getClassName(lines),
          fields: [],
        };
        let documentation = '';
        for (let i = 1; i + 1 < lines.length; i++) {
          const line = this.lineArray(lines[i]);
          if (line[0].includes('//')) {
            documentation = documentation ? documentation + '\n' + line.join(' ') : line.join(' ');
          } else if (line[0].includes('@@')) {
            modelObject.map = this.getMap(lines[i]);
          } else {
            const type = this.getType(line);
            const field: Field = {
              name: line[0],
              type,
              isId: !!line.find((part) => part.startsWith('@id')),
              unique: !!line.find((part) => part.startsWith('@unique')),
              defaultValue: line.find((part) => part.startsWith('@default')),
              list: line[1].includes('[]'),
              required: !line[1].includes('[]') && !line[1].includes('?'),
              kind: this.getKind(type),
              documentation,
              map: this.getMap(lines[i]),
            };

            if (field.kind === 'object') {
              field.relation = this.getRelation(lines[i]);
            }
            modelObject.fields.push(field);
            documentation = '';
          }
        }
        modelObject.documentation = this.getModelDocumentation(modelObject.name);
        modelObject.fields
          .filter((item) => item.kind !== 'object')
          .forEach((item) => {
            let relationField = false;
            modelObject.fields
              .filter((field) => field.kind === 'object')
              .forEach((field) => {
                if (!relationField) {
                  relationField = !!field.relation?.fields?.includes(item.name);
                }
              });
            item.relationField = relationField;
          });
        this.schemaObject.models.push({ ...modelObject });
      }
    }
  }

  private getEnums() {
    if (this.enums) {
      for (const item of this.enums) {
        const lines = this.blockLines(item);
        const itemObject: Enums = {
          name: this.getClassName(lines),
          fields: [],
        };
        for (let i = 1; i + 1 < lines.length; i++) {
          const line = this.lineArray(lines[i]);
          !line[0].includes('//') && itemObject.fields.push(line[0]);
        }
        this.schemaObject.enums.push({ ...itemObject });
      }
    }
  }
}
