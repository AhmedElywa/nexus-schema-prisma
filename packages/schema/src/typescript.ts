import { Field, SchemaObject } from '@paljs/types';
import { ConvertSchemaToObject } from './json';

export class GenerateTypeScript {
  private schema: SchemaObject;
  private code: string[] = [];
  private scalar: Record<string, string> = {
    Int: 'number',
    Float: 'number',
    Decimal: 'number',
    BigInt: 'number',
    String: 'string',
    Boolean: 'boolean',
    DateTime: 'Date',
    Json: 'any',
  };
  constructor(path: string) {
    this.schema = new ConvertSchemaToObject(path).run();
  }

  run() {
    // generate models
    const models: string[] = [];
    this.schema.models.forEach((model) => {
      const fields: string[] = [`export interface ${model.name} {`];
      model.fields.forEach((field) => {
        fields.push(`${field.name}${field.required ? '' : '?'}: ${this.getType(field)}`);
      });
      fields.push('}');
      models.push(fields.join('\n'));
    });
    this.code.push(models.join('\n\n'));

    // generate enums
    const enumsTypes: string[] = [];
    this.schema.enums.forEach((item) => {
      const values: string[] = [`export enum ${item.name} {`];
      item.fields.forEach((item2) => {
        values.push(`${item2} = "${item2}",`);
      });
      values.push('}');
      enumsTypes.push(values.join('\n'));
    });
    this.code.push(enumsTypes.join('\n\n'));

    return this.code.join('\n\n');
  }

  private getType(field: Field) {
    switch (field.kind) {
      case 'scalar':
        return `${this.scalar[field.type]}${field.list ? '[]' : ''}`;
      default:
        return `${field.type}${field.list ? '[]' : ''}`;
    }
  }
}
