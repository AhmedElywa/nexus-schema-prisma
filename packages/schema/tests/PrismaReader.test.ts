import { PrismaReader } from '../src';
import { join } from 'path';

describe('test the Prisma Reader class', () => {
  const prismaReader = new PrismaReader(join(__dirname, './schemas/schema.prisma'));

  test('read prisma models', () => {
    expect(prismaReader.models).toMatchSnapshot();
  });

  test('read prisma enums', () => {
    expect(prismaReader.enums).toMatchSnapshot();
  });

  test('read prisma model documentation', () => {
    expect(prismaReader.getModelDocumentation('User')).toMatchSnapshot();
  });
});
