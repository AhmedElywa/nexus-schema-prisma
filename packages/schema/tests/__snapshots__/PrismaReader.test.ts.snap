// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test the Prisma Reader class read prisma enums 1`] = `
Array [
  "
enum Role {
  ADMIN
  USER
}
",
]
`;

exports[`test the Prisma Reader class read prisma model documentation 1`] = `"// User modal documentation"`;

exports[`test the Prisma Reader class read prisma models 1`] = `
Array [
  "
model User {
  id          Int      @id @default(autoincrement())
  createdAt   DateTime @default(now())
  email       String   @unique
  name        String?
  // password documentation
  // password second line documentation
  password    String
  permissions Json     @default(\\"{}\\")
  posts       Post[]
}
",
  "
model Post {
  id        Int      @id @default(autoincrement())
  published Boolean  @default(false)
  title     String
  author    User?    @relation(fields: [authorId], references: [id])
  authorId  Int?
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
}
",
]
`;
